source/llvm/ASTNodeCodeGen.cpp

```diff
@@ -637,6 +639,159 @@ llvm::Value* ASTNodeCodeGen::delayExprCodeGen(const libsbml::ASTNode* ast)
+    //ModelDataIRBuilder mdbuilder(modelData, dataSymbols, builder);
+    Value* rate = NULL;
+    resolver.recursiveSymbolPush("rateOf(" + name + ")");
+    //Looking for a rate.  Our options are: local parameter, floating species, rate rule target, assignment rule target (an error), and everything else is zero.
+    if (resolver.isLocalParameter(name))
+    {
+        rate = ConstantFP::get(builder.getContext(), APFloat(0.0));
+    }
+    else if (dataSymbols.isIndependentFloatingSpecies(name))
+    {
+        //NOTE:  if we stored rates persistently, we could use the following instead:
+        //rate = mdbuilder.createFloatSpeciesAmtRateLoad(name, name + "_amtRate");
+        const Model* model = ctx.getModel();
+        const Species* species = model->getSpecies(name);
+        bool speciesIsAmount = species->getHasOnlySubstanceUnits();
+        string compId = species->getCompartment();
+        bool compIsConst = true;
+
+        if (!speciesIsAmount) {
+            //Determine if the compartment changes in time.
+            const Rule* compRule = model->getRule(compId);
+            if (compRule != NULL) {
+                if (compRule->getTypeCode() == SBML_ASSIGNMENT_RULE) {
+                    std::stringstream err;
+                    err << "Unable to calculate rateOf(" << name << "), because "
+                        << name << " is a concentration, not an amount, but its compartment ("
+                        << compId << ") chages due to an assignment rule.  Since we cannot"
+                        << " calculate derivatives on the fly, we cannot determine the rate of"
+                        << " change of the species " << name << " concentration";
+                    throw_llvm_exception(err.str());
+                }
+                else {
+                    //The only other option is a rate rule, which means the compartment changes in time.
+                    compIsConst = false;
+                }
+            }
+        }
+
+        //The functions are different depending on how complicated things get, so we need references.  The final Value will be created from the overallRef ASTNode.
+        ASTNode amtRate(AST_PLUS);
+        ASTNode* amtRateRef = &amtRate;
+        ASTNode* overallRef = &amtRate;
+
+        //Deal with conversion factors:
+        string conversion_factor = "";
+        if (model->isSetConversionFactor()) {
+            conversion_factor = model->getConversionFactor();
+        }
+        if (species->isSetConversionFactor()) {
+            conversion_factor = species->getConversionFactor();
+        }
+        if (!conversion_factor.empty()) {
+            amtRate.setType(AST_TIMES);
+            ASTNode* cf = new ASTNode(AST_NAME);
+            ASTNode* unscaledRate = new ASTNode(AST_PLUS);
+            cf->setName(conversion_factor.c_str());
+            amtRate.addChild(unscaledRate);
+            amtRate.addChild(cf);
+            amtRateRef = unscaledRate;
```

Comment:
A little reordering of these lines would help to make it more readable.

Reply:
Done!